crates/bevy_ui/src/flex/convert.rs

```diff
@@ -1,99 +1,77 @@
-use taffy::style::LengthPercentageAuto;
+use bevy_math::Vec2;
 
 use crate::{
     AlignContent, AlignItems, AlignSelf, Display, FlexDirection, FlexWrap, JustifyContent,
     PositionType, Size, Style, UiRect, Val,
 };
 
 impl Val {
-    fn scaled(self, scale_factor: f64) -> Self {
+    fn scaled(self, scale_factor: f64, physical_size: Vec2) -> Self {
         match self {
             Val::Auto => Val::Auto,
             Val::Percent(value) => Val::Percent(value),
             Val::Px(value) => Val::Px((scale_factor * value as f64) as f32),
+            Val::VMin(value) => Val::Px(physical_size.x.min(physical_size.y) * value / 100.),
+            Val::VMax(value) => Val::Px(physical_size.x.max(physical_size.y) * value / 100.),
+            Val::Vw(value) => Val::Px(physical_size.x * value / 100.),
+            Val::Vh(value) => Val::Px(physical_size.y * value / 100.),
         }
     }
-
-    fn to_inset(self) -> LengthPercentageAuto {
-        match self {
-            Val::Auto => taffy::style::LengthPercentageAuto::Auto,
-            Val::Percent(value) => taffy::style::LengthPercentageAuto::Percent(value / 100.0),
-            Val::Px(value) => taffy::style::LengthPercentageAuto::Points(value),
-        }
-    }
-}
-
-impl UiRect {
-    fn scaled(self, scale_factor: f64) -> Self {
-        Self {
-            left: self.left.scaled(scale_factor),
-            right: self.right.scaled(scale_factor),
-            top: self.top.scaled(scale_factor),
-            bottom: self.bottom.scaled(scale_factor),
-        }
-    }
-}
-
-impl Size {
-    fn scaled(self, scale_factor: f64) -> Self {
-        Self {
-            width: self.width.scaled(scale_factor),
-            height: self.height.scaled(scale_factor),
-        }
-    }
-}
-
-impl<T: From<Val>> From<UiRect> for taffy::prelude::Rect<T> {
-    fn from(value: UiRect) -> Self {
-        Self {
-            left: value.left.into(),
-            right: value.right.into(),
-            top: value.top.into(),
-            bottom: value.bottom.into(),
-        }
-    }
-}
-
-impl<T: From<Val>> From<Size> for taffy::prelude::Size<T> {
-    fn from(value: Size) -> Self {
-        Self {
-            width: value.width.into(),
-            height: value.height.into(),
-        }
-    }
-}
-
-impl From<Val> for taffy::style::Dimension {
-    fn from(value: Val) -> Self {
-        match value {
+    fn into_dimension(self, scale_factor: f64, physical_size: Vec2) -> taffy::style::Dimension {
+        match self.scaled(scale_factor, physical_size) {
             Val::Auto => taffy::style::Dimension::Auto,
             Val::Percent(value) => taffy::style::Dimension::Percent(value / 100.0),
             Val::Px(value) => taffy::style::Dimension::Points(value),
+            _ => unreachable!(),
```

Comment:
I think both the `scale_factor` and `physical_size` conversions should be implemented here and in the other `into_*` functions rather than having `unreachable!()`. That means that `Val` is always in logical pixels which keeps things simple. It means a bit of repetition, but I think that's acceptable.

Alternatively a *private* `PhysicalVal` (not exposed to users) could make sense. But it might just be easier to use Taffy types for physical units?

Reply:
Scaled returning a `Dimension`, seems easiest I think.

Reply:
or rather rename `scaled` to `into_dimension`